### PR TITLE
feature: implement friend mechanics through a follower/following system 

### LINKED ISSUE #80: Friends: update users MVVM to include friends
Add a friend field to users (e.g so a user can see all their friends' notes in the overview). Add mechanics to add, remove and list all friends. Add tests.

### PR DESCRIPTION
**Pull request by robertaleG**
# Description

I have implemented a way to have friends in the app using public asymmetrical _follower-following_ relationships: a user can choose to follow some other user without them having to follow the user back.
This is important for content creation/sharing platforms where few users produce quality content which all other users wish to have access to.

To model this, I have given the `User` an attribute _friends_ whose type is a custom class `Friends(following : List<String>, followers : List<String>)` containing a List of follower ids and a List of following ids. This is preferred over using a List<User> which could take a lot more space than a simple List<String> once we start adding more fields to `User`.    
The IDs are the UIDs of the users in the Firestore database and are used to quickly index and query users.

NOTE: For now, every profile is public and there are no "follow requests" typical in some social networks.

Fixes #80 

# How Has This Been Tested?

I have added completely reworked and modularised the unit tests for the UserRepositoryFirestore and UserViewModel, as we will need to add even more tests to those files later with the upcoming features.
I then added unit tests for all the new functions I coded, and updated the androidTests to still pass.

I have tried the code on Firestore by adding custom buttons to follow/unfollow the UIDs of other existing users. It works perfectly.

### Unit Tests

As said above, I reworked `UserViewModelTest` and `UserRepositoryFirestoreTest`, and updated `UserTest`. 
  
### UI Tests

I updated `ProfileScreenTest` because of the new refreshUser functions which affected earlier code and tests.

### End to end Tests

None added.

  
# Dependencies

No new dependencies added.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt [hunk 1]
```diff
@@ -19,15 +19,15 @@ class UserViewModel(private val repository: UserRepository) : ViewModel() {
...
 
+  /** Exception thrown when a user is not logged in. */
+  class UserNotLoggedInException : Exception("User Not Logged In")
+
   /** Initializes the UserViewModel and the repository. */
   init {
     repository.init(FirebaseAuth.getInstance()) {}
   }
 
-  fun setCurrentUser(user: User) {
```
[the remark this answers is not in the document]
Comment (LawrenceLeitgib) [reply app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt:27]: solved

File: app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt [hunk 2]
```diff
@@ -48,25 +48,52 @@ class UserViewModel(private val repository: UserRepository) : ViewModel() {
...
    * @param user The user to add.
    * @param onSuccess Callback to be invoked when the addition is successful.
    * @param onFailure Callback to be invoked if an error occurs.
    */
   fun addUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
-    repository.addUser(user, onSuccess, onFailure)
+    repository.addUser(
+        user,
+        {
+          _currentUser.value = user
```
[the remark this answers is not in the document]
Comment (LawrenceLeitgib) [reply app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt:76]: solved

File: app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt [hunk 3]
```diff
@@ -48,25 +48,52 @@ class UserViewModel(private val repository: UserRepository) : ViewModel() {
...
+          onSuccess()
+        },
+        onFailure)
   }
 
   /**
-   * Updates the information of an existing user.
+   * Updates the information of the current user locally and in the database.
```
[the remark this answers is not in the document]
Comment (LawrenceLeitgib) [reply app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt:83]: solved

File: app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt [hunk 4]
```diff
@@ -48,25 +48,52 @@ class UserViewModel(private val repository: UserRepository) : ViewModel() {
...
    *
    * @param user The user with updated information.
    * @param onSuccess Callback to be invoked when the update is successful.
    * @param onFailure Callback to be invoked if an error occurs.
    */
   fun updateUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
-    repository.updateUser(user, onSuccess, onFailure)
```
[the remark this answers is not in the document]
Comment (LawrenceLeitgib) [reply app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt:69]: solved

File: app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt [hunk 5]
```diff
@@ -75,25 +102,32 @@ class UserViewModel(private val repository: UserRepository) : ViewModel() {
...
       email: String,
       onSuccess: (User) -> Unit,
       onUserNotFound: () -> Unit,
       onFailure: (Exception) -> Unit
   ) {
-    repository.getUserByEmail(email, onSuccess, onUserNotFound, onFailure)
+    repository.getUserByEmail(
+        email,
+        { usr ->
+          _currentUser.value = usr
```
[the remark this answers is not in the document]
Comment (LawrenceLeitgib) [reply app/src/main/java/com/github/onlynotesswent/model/users/UserViewModel.kt:126]: solved

File: app/src/test/java/com/github/onlynotesswent/model/users/UserRepositoryFirestoreTest.kt [hunk 6]
```diff
@@ -400,34 +312,94 @@ class UserRepositoryFirestoreTest {
...
+    // Call getUsersById method
     var users: List<User>? = null
-    userRepositoryFirestore.getUsers({ users = it }, { assert(false) })
-    verify(mockCollectionReference, timeout(1000)).get()
+    userRepositoryFirestore.getUsersById(listOf(user.uid), { users = it }, { assert(false) })
+
+    // Verify if Firestore collection was called
+    verify(mockCollectionReference, timeout(1000)).whereIn("uid", listOf(user.uid))
+
+    // Assertions to verify that the correct user was returned
```
Comment (LawrenceLeitgib) [inline app/src/test/java/com/github/onlynotesswent/model/users/UserRepositoryFirestoreTest.kt:402]: Good tests and good refractor of old tests !

**Review by LawrenceLeitgib — APPROVED** [review]
Overall good code, a lot of test but this is good. Juste a few questions before merging but execpt that all good.

File: app/src/androidTest/java/com/github/onlynotesswent/ui/user/ProfileScreenTest.kt [hunk 7]
```diff
@@ -115,4 +122,32 @@ class ProfileScreenTest {
...
+
+    composeTestRule.onNodeWithTag("saveButton").performClick() // error occurs
+    composeTestRule.onNodeWithTag("inputUserName").assert(hasError()) // error shown
+
+    assert(userViewModel.currentUser.value?.userName == "testUserName") // did not change
+    assert(userViewModel.currentUser.value?.firstName == "testFirstName") // did not change
+  }
+
+  @Test
+  fun saveButtonDisabledWhenUserNameIsEmpty() {
```
[the remark this answers is not in the document]
Comment (robertaleG) [reply app/src/androidTest/java/com/github/onlynotesswent/ui/user/ProfileScreenTest.kt:144]: These features were not tested, I completed the old tests.

File: app/src/main/java/com/github/onlynotesswent/model/users/User.kt [hunk 8]
```diff
@@ -20,5 +20,18 @@ data class User(
...
+/**
+ * Data class representing a user's friends (following and followers), represented by lists of UIDs.
+ *
+ * @constructor Creates a Friends object with empty lists.
+ * @property following The list of users that the user is following.
+ * @property followers The list of users that are following the user.
+ */
+data class Friends(
+    val following: List<String> = emptyList(),
+    val followers: List<String> = emptyList()
```
[the remark this answers is not in the document]
Comment (robertaleG) [reply app/src/main/java/com/github/onlynotesswent/model/users/User.kt:36]: I meant to say that we don't need to have a list of Users but just a list of user ids, which we can then fetch from the database when needed. That way on the database there is no circular definition issues or duplicate users.

**Pull request merged**